onnxruntime/core/providers/cpu/tensor/upsample_antialias.h

```diff
@@ -0,0 +1,763 @@
+// Copyright c Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <algorithm>
+#include <cmath>  // for round
+#include <vector>
+#include "core/framework/tensor.h"
+#include "gsl/span"
+#ifndef SHARED_PROVIDER
+#include "core/framework/op_kernel.h"
+#endif
+#include "core/providers/cpu/tensor/upsamplebase.h"
+
+namespace onnxruntime {
+
+namespace ConstValue {
+constexpr int32_t mag_factor = 1 << (22 - 1);
+}
+
+template <typename T>
+struct FilterParamsBaseAntiAlias {
+  std::vector<int64_t> bound;
+  std::vector<int64_t> out_of_bound_idx;
+  int64_t window_size = 2;
+  IAllocatorUniquePtr<T> weight_coefficients;
+};
+
+template <typename T>
+struct FilterParamsAntiAlias {
+  float support_size = 2.0f;
+  float cubic_coeff_a = -0.75f;
+
+  /* Handles values form -640 to 639. */
+  uint8_t* clip8_lookups_table{nullptr};
+
+  FilterParamsBaseAntiAlias<T> dim_x;
+  FilterParamsBaseAntiAlias<T> dim_y;
+  FilterParamsBaseAntiAlias<T> dim_z;
+
+  void init_clip_lookup() {
+    // if we have already initialized the lookup table, just return
+    // ideally we could have a global lookup table, but that account for too much space.
+    if (clip8_lookups_table[1279] == 255) {
+      return;
+    }
+
+    // taken from https://github.com/python-pillow/Pillow/blob/66add095a50d76c35c7f58643461f2edf78a3f05/src/libImaging/Resample.c#L94
+    //  we need to handle negative values
+    //  it's equivalent to :x = np.clip(x, 0, 255) where x \in [-640, 639]
+    // we will accept a negative x for (&clip8_lookups_table[640])[x] means clip8_lookups_table +640 -x
+    for (int i = 0; i < 1280; ++i) {
+      clip8_lookups_table[i] = static_cast<uint8_t>(std::min(std::max(i - 640, 0), 255));
+    }
+  }
+  virtual ~FilterParamsAntiAlias() = default;
+  virtual float filter(float x) const = 0;
+};
+
+template <typename T>
+struct BilinearParamsAntiAlias : FilterParamsAntiAlias<T> {
+  // taken from
+  // https://github.com/python-pillow/Pillow/blob/6812205f18ca4ef54372e87e1a13ce4a859434df/
```

Comment:
Need to add python-pillow to cgmanifest.json and ThirdPartyNotices.txt

Reply:
Just wondering, why we need to add it into " cgmanifest.json"? The implementation doesn't directly depend on python-pillow.

Reply:
But we're copying code from there that will be included in the shipped product so AFAIK it needs to be added as the usage of it won't be automatically discoverable like using it as say a submodule.

Reply:
I see. Added as a git repo.